respy/simulate.py

```diff
@@ -159,43 +146,47 @@ def simulate(
     state_space = solve_with_backward_induction(state_space, optim_paras, options)
 
     # Prepare simulation.
+    n_simulation_periods = int(df.index.get_level_values("period").max() + 1)
+
+    # Prepare shocks.
     n_wages = len(optim_paras["choices_w_wage"])
     base_draws_sim_transformed = transform_base_draws_with_cholesky_factor(
         base_draws_sim, optim_paras["shocks_cholesky"], n_wages
     )
     base_draws_wage_transformed = np.exp(base_draws_wage * optim_paras["meas_error"])
 
-    df = df.copy()
+    # Store the shocks inside the DataFrame. The sorting ensures that regression tests
+    # still work.
+    df = df.sort_index(level=["period", "identifier"])
+    for i, choice in enumerate(optim_paras["choices"]):
+        df[f"shock_reward_{choice}"] = base_draws_sim_transformed[:, i]
+        df[f"meas_error_wage_{choice}"] = base_draws_wage_transformed[:, i]
+    df = df.sort_index(level=["identifier", "period"])
+
     df = _extend_data_with_sampled_characteristics(df, optim_paras, options)
-    is_n_step_ahead = df.index.get_level_values("identifier").duplicated().sum() == 0
 
-    # Start simulating.
-    data = []
+    core_columns = create_core_state_space_columns(optim_paras)
+    is_n_step_ahead = np.any(df[core_columns].isna())
+
     for period in range(n_simulation_periods):
 
         # If it is a one-step-ahead simulation, we pick rows from the panel data. For
-        # n-step-ahead simulation, ``df`` always contains only data of the current
-        # period.
-        current_states = df.query("period == @period").to_numpy(dtype=np.uint32)
-
-        rows = _simulate_single_period(
-            period,
-            current_states,
-            state_space,
-            base_draws_sim_transformed,
-            base_draws_wage_transformed,
-            optim_paras,
+        # n-step-ahead simulation, `df` always contains only data of the current period.
+        current_df = df.query("period == @period").copy()
+
+        current_df_extended = _simulate_single_period(
+            current_df, state_space, optim_paras
         )
 
-        data.append(rows)
+        # Add all columns with simulated information to the complete DataFrame.
+        df = df.reindex(columns=current_df_extended.columns) if period == 0 else df
+        df = df.combine_first(current_df_extended)
 
-        if is_n_step_ahead:
-            choices = rows[:, 1].astype(np.uint8)
-            df = _apply_law_of_motion(df, choices, optim_paras)
+        if is_n_step_ahead and period != n_simulation_periods - 1:
+            next_df = _apply_law_of_motion(current_df_extended, optim_paras)
+            df = df.combine_first(next_df)
```

Comment:
I think this idea is very cool! That makes everything more coherent!

Reply:
Yes :) I think it was you who told me to use a single DataFrame while working on #277, but at the time I thought it was not necessary. Now, I see how convenient the DataFrame is, even more for parallelizing the simulation in #310.